Implement schema migration and panic when trying to downgrade storage

serathius opened the pull request:
Part of milestone 1 of https://github.com/etcd-io/etcd/issues/13168
cc @ptabor

lilic commented on server/storage/schema/schema.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+var (
+	V3_5 = semver.Version{Major: 3, Minor: 5}
```
Since we already have these and to me they seem more like constants, what do you think we take them from https://github.com/etcd-io/etcd/blob/f4fad92e0d0240d1495cc6bf9d2f4d1fbeab4e4b/server/etcdserver/version/version.go#L28-L29

lilic commented on server/storage/schema/schema.go:
```diff
@@ -0,0 +1,135 @@
+		return fmt.Errorf("cannot determine storage version: %w", err)
+	}
+	if ver.Major != target.Major {
+		lg.Panic("Chaning major storage version is not supported",
```
```suggestion
		lg.Panic("Changing major storage version is not supported",
```

serathius replied:
Fixed.

lilic commented on server/storage/schema/schema.go:
```diff
@@ -0,0 +1,135 @@
+		)
+	}
+	if ver.Minor > target.Minor {
+		lg.Panic("Downgrades are not yet supported",
```
Should we panic or return error maybe instead?

Maybe: 

```suggestion
		lg.Panic("Target version is lower than the current version, specific higher version, downgrades are not yet supported",
```

serathius replied:
This is one breaking change we are introducing, future etcd version (for example v3.7) cannot be passed to v3.6 without doing the downgrade process.

lilic commented on server/storage/schema/schema.go:
```diff
@@ -0,0 +1,135 @@
+	if term == 0 {
+		return nil, fmt.Errorf("missing %q key", MetaTermKeyName)
+	}
+	copied := V3_5
```
Can you maybe explain this a bit? A comment for future might be useful as well. Thanks!

serathius replied:
This is to avoid returning a pointer to global variable, we copy the value locally and return pointer to local copy. Preferred to use a variable name then adding comment that will get outdated. This is something also done in other parts of code, so I didn't think this is something unusual.

lilic commented on server/storage/schema/schema.go:
```diff
@@ -0,0 +1,135 @@
+		}
+	}
+	// Storage version is available since v3.6
+	if next != V3_5 {
```
What if its 3.4 or smaller?

serathius replied:
Current detectStorageVersion should never return versions < v3.5, but you're right this function should not assume that.

lilic replied:
Yes exactly, people tend to reuse functions :D

lilic commented on server/storage/schema/schema_test.go:
```diff
@@ -0,0 +1,153 @@
+			expectPanic:     true,
+		},
+		{
+			name:            "Upgrade to v3.7 is not supported",
```
Lets add a test for lower than 3.5 as well, for example 3.4.

serathius replied:
Done.

lilic commented on server/storage/schema/confstate_test.go:
```diff
@@ -12,14 +12,14 @@
 // limitations under the License.
 
-package buckets
+package schema
```
Since this is a public package and not internal would this be considered breaking change? I know some projects import etcd for example.

serathius replied:
Is there any official policy what package have stable functions/interface? It makes sense for libraries, but it's strange for me that we would have something like that for server code. Do we care if someone imports a package that we haven't officially commit to stabilize?

 I mean that in a way that maybe we should start an effort to decide what packages are stable and which are definitely not.

Reverted this change.

lilic replied:
Agreed, I don't think there is an official policy, but its not part of an `internal` directory. But might be nice to discuss this in our next meeting, what do you think?

serathius replied:
SG

lilic reviewed (commented):
Couple of comments and suggestions 🎉 

I would find it useful to maybe add a bit more context to the description of the PRs in the future, as its harder to follow if I miss one or two. I read the design doc and the issue but still might be useful IMHO, thank you!

serathius commented:
PTAL @lilic

lilic commented on server/etcdserver/server.go:
```diff
@@ -2414,8 +2423,10 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
-		s.storageVersionUpdated.Do(func() {
-			err := serverversion.UpdateStorageVersion(s.lg, s.be.BatchTx())
+		// Update storage schema after snapshot as fields introduced in v3.5 should be set.
+		// Remove in v3.7
```
Nit and could we open an issue with 3.7 milestone label so we don't forget?

```suggestion
		// TODO: Remove in v3.7
```

lilic commented on server/etcdserver/version/version.go:
```diff
@@ -29,27 +29,39 @@ var (
+		)
+	}
+	if ver.Minor > target.Minor {
+		lg.Panic("Target version is lower than the current version, specific higher version, downgrades are not yet supported",
```
```suggestion
		lg.Panic("Target version is lower than the current version, specificy higher version, downgrades are not yet supported",
```

serathius replied:
```suggestion
		lg.Panic("Target version is lower than the current version, downgrades are not yet supported, please specify higher version",
```

lilic commented on server/etcdserver/version/version.go:
```diff
@@ -29,27 +29,39 @@ var (
-	default:
-		lg.Warn("unknown storage version", zap.String("storage-version", v.String()))
+	if ver.Major != target.Major {
+		lg.Panic("Changing major storage version is not supported",
```
Just a thought, what will happen in v4, how will downgrades/migrates work then?

serathius replied:
We will update this code when we decide to release v4.0.0. If for example v3.7 will be the last v3.X release then we will add a code in v4.0.0 that will treat v3.7 as predecessor. In that situation migration capabilities should be:
* etcd v3.6 can run on data from: v3.5, v3.6
* etcdutl v3.6 can migrate storage between versions: v3.5, v3.6
* etcd v3.7 can run on data from: v3.6, v3.7
* etcdutl v3.7 can migrate storage between versions: v3.5, v3.6, v3.7
* etcd v4.0 can run on data from: v3.7, v4.0
* etcdutl v4.0 can migrate storage between versions: v3.5, v3.6, v3.7, v4.0

In short etcd will work with data from it's or previous release. However if you have etcdutl from latest release you should be able to upgrade and downgrade storage freely.

lilic replied:
Thank you for the details!

lilic commented on server/etcdserver/version/version.go:
```diff
@@ -65,3 +77,80 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+	} else {
+		target = semver.Version{Major: current.Major, Minor: current.Minor - 1}
+	}
+	ms := migrations(lg, current, target)
```
Nit: maybe instead lets name this newMigration, wdyt?
```suggestion
	ms := newMigration(lg, current, target)
```

serathius replied:
This functions returns list of boltdb modifications "migrations" that need to be executed. Changing function name to `newMigration` implies that single object is returned. We could add a new structure, but then we would need to change the name of interface.

lilic replied:
maybe `getMigrations`? Just a thought. More of a nit really.

serathius replied:
Renamed `migration` to `migtionAction` and introduced a `newMigration` function like you proposed. I think it's more clearer as we don't reuse same name. 

`migration` struct is a change between two neighboring versions (v3.5 and v3.6). `Migration` is a collection of actions that implement `migrationAction` interface. Action is a unit of change like add a new field.

lilic commented on server/etcdserver/version/version.go:
```diff
@@ -65,3 +77,80 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+func (m *newField) UnsafeUpgrade(tx backend.BatchTx) error {
+	tx.UnsafePut(m.bucket, m.fieldName, m.fieldValue)
+	return nil
+}
```
New line missing after this.

serathius replied:
Was sad that gofmt doesn't fix this.

lilic replied:
Yeah surprised by that as well.

lilic commented on server/etcdserver/version/version.go:
```diff
@@ -65,3 +77,80 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+	UnsafeDowngrade(backend.BatchTx) error
+}
+
+type newField struct {
```
Bit unsure what this struct name conveys here but can't think of a better suggestion. 🤔  Maybe just `field`?

serathius replied:
This struct doesn't represent a field, but an action of adding a field. Would name `addNewField` be better?

lilic replied:
Sounds good to me 👍

lilic commented on server/etcdserver/version/version_test.go:
```diff
@@ -20,61 +20,109 @@ import (
+	V3_7 = semver.Version{Major: 3, Minor: 7}
+)
+
 func TestUpdateStorageVersion(t *testing.T) {
```
Does this test still belong in the version test. As we don't really call anything from that file? Just a thought.

serathius replied:
Forgot to update the test name, it should be `TestMigrate`. It tests the `Migrate` function from `version.go`, so it should stay here. However I plan to move it to `storage/schema/schema.go` when https://github.com/etcd-io/etcd/pull/13198 is merged.

lilic replied:
Nice, sgtm 👍

lilic reviewed (commented):
Couple more comments. I do believe we need more consensus on the breaking change here.

cc @hexfusion

serathius commented:
Addressed all issues, PTAL

serathius commented:
Send an announcement about the breaking change to make sure community is aware of it. https://groups.google.com/g/etcd-dev/c/-u_vl_G-nRg

ptabor commented on server/etcdserver/server.go:
```diff
@@ -2136,10 +2136,10 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
+		s.migrateStorageSchema.Do(func() {
+			// Update storage schema after snapshot as setting it requires all fields introduced in v3.5 to be set. Those field can only be set after snapshot.
+			// TODO: Remove in v3.7 as we should be able to depend on running Migrate during etcdserver boostrap.
+			err := schema.Migrate(s.lg, s.be.BatchTx(), schema.V3_6)
```
Shell we assert here based on:
  https://github.com/etcd-io/etcd/blob/main/api/version/version.go
  
  As this code overlooked in 3.7 upgrade process might be dangerous (with hardcoded to 3.6 migration)

serathius replied:
Done

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -28,27 +28,39 @@ var (
-	defer tx.Unlock()
-	v, err := detectStorageVersion(lg, tx)
+// Migrate updates storage version to provided target version.
+func Migrate(lg *zap.Logger, tx backend.BatchTx, target semver.Version) error {
```
nit: maybe worth to put it in separate migrate.go as it's going to become huge on its own...

serathius replied:
Ack, when defining schema, will split it from migration.

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -28,27 +28,39 @@ var (
+			return err
+		}
+		ver = next
+		lg.Info("upgraded storage version", zap.String("storage-version", ver.String()))
```
new-storage-version (to ensure its post-migration state)

serathius replied:
Done

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -28,27 +28,39 @@ var (
+		)
+	}
+	if ver.Minor > target.Minor {
+		lg.Panic("Target version is lower than the current version, downgrades are not yet supported, please specify higher version",
```
This should contain recommendation to use etcdutl, right ? 
[not yet supported, but maybe worth adding TODO]

serathius replied:
Migrate function will also be used by etcdutl, will leave a TODO.

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -64,3 +76,92 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+
+var versionMigateActions = map[semver.Version][]migrateAction{
+	V3_6: {
+		&addNewFieldAction{bucket: Meta, fieldName: MetaStorageVersionName, fieldValue: []byte("")},
```
Why putting an empty "" value (instead of real value) is a proper upgrade path ? 
Probably requires a comment at list.

serathius replied:
This is a special case, normally it should not be empty.
I created an action to so the field will be handled by migrator (removed during downgrade).

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -64,3 +76,92 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+	upgrade := m.current.LessThan(m.target)
+	tx.Lock()
+	defer tx.Unlock()
+	for _, m := range m.actions {
```
I wonder whether downgrades actions should be executed from the end...

Currently its a single transaction so it doesn't matter, but if it would mutate wal or storege v1 (yes: deprecated), it might have some reasons.   No strong use-case... just for consideration e.g. in the future.

serathius replied:
Done

ptabor commented on server/storage/schema/schema.go:
```diff
@@ -64,3 +87,93 @@ func detectStorageVersion(lg *zap.Logger, tx backend.ReadTx) (*semver.Version, e
+
+var versionMigateActions = map[semver.Version][]migrateAction{
+	V3_6: {
+		&addNewFieldAction{bucket: Meta, fieldName: MetaStorageVersionName, fieldValue: []byte("") /* leaving empty as StorageVersion should be set by migrator */},
```
How about making it a dedicated enum, such it's easy to find all places in the code that take care of providing the actual value instead of ""

serathius replied:
Makes sense, added a variable.

ptabor commented on CHANGELOG-3.6.md:
```diff
@@ -8,6 +8,10 @@ Previous change logs can be found at [CHANGELOG-3.5](https://github.com/etcd-io/
 
+### Breaking Changes
+
+- `etcd` will no longer start on data dir created by newer versions (for example etcd v3.6 will not run on v3.7+ data dir). To downgrade data dir please check out `etcdutl migrate` command.
```
> will no longer start on data dir created by newer versions (for example etcd v3.6 will not run on v3.7+ data dir).

Would we need etcdutl 3.7 to perform the downgrade or etcdutl-3.6 would handle this ? 

Thinking about scenario: 
  - we attempted upgrade of clusters.  It failed. Some logic decided to restart from previous version of 'image'. 
  - the new image contains etcdutl-3.6 and etcd 3.6
  - the image cannot contain self-healing logic. 
 
TBH: A single minor version downgrades are pretty important from reliability perspective.

serathius replied:
This is a good point, the current storage versioning proposes a solution, but it might be worth discussing more details here.

When talking about storage downgrades it's important to look in context of whole downgrade process. As etcd WAL can contain new etcd entries, etcd downgrade cannot be done ad hock. It requires a whole procedure to ensure that all members are aware of downgrade and prepare for it. 

Skipping this procedure and just starting older etcd on the same data will always come with a risk encountering unknown WAL entry type. The restriction for downgrading is introduced to guide users towards the safer process as default. Still setup you described is pretty common in kubernetes and we should support it.

By default downgrading v3.7 will require using etcdutl v3.7, this way the tool with full information can make sure that the downgrade is correct and safe, still downgrading using older version is possible, but not as safe. This is why `etcdutl migrate --force` was introduced. In this mode older etcdutl will be can downgrade newer version, based on assumption that etcd bbolt schema is backward compatible. Additionally it should purge all the unknown fields and WAL entries to make. As this will strongly depend on backward compatibility between one minor version, we should add dedicated testing that we will not break it by mistake.

Does using `etcdutl migrate --force` in this scenario address your concern?

ptabor replied:
The previous experience was ~safe  because of 2-release. In release N+1 we added support for feature and 'experimental flags' activating it, but by default we were not writing WAL logs that couldn't be interpreted by N release. 

I agree that N+2 releases is killing velocity. I'm thinking whether we can divide lifecycle of the N+1 server into 2 phases: 
   - pre cluster upgrade (no new features are being used)
   - post cluster upgrade (quorum agreed to take new version) when new WAL logs are accepted.
 I think it's how it was described in the design. 

So maybe the Changelog phrasing that when quorum commits cluster upgrade than explicit downgrade is being needed. 

This is big change and we need @gyuho approval.

serathius commented:
Talked about this change with @ptabor, there is one error in the current PR. Setting storage version should not be set during etcd init, but when cluster version is changed. As described in the design storage version should follow cluster version, delayed only by ensuring that written data is backward compatible.

This means that this PR doesn't introduce as impactful breaking change as originally discussed. Restrictions introduced there will be very similar to already existing restrictions caused by cluster version. When cluster is bootstraped it first agrees on cluster version based on quorum. It happens within seconds from cluster init, and after it is impossible to downgrade cluster version, which is very similar to storage version restriction proposed here.

How storage version should be used:
* When bootstraping from clean state, storage version shoudl be set to etcd binary version. 
* When bootstrapping with existing storage, etcdserver should validate if it supports it, but don't do any migrations.
* If storage version differs from cluster version, we should try to migrate storage (might require starting a background process to wait till snapshot)

Setting storage version asynchronously by members (without quorum) should not be a problem, as it should only impact data format and not the information stored (would be good to have tests for that). Only problem I see is how to store new keys, during upgrade time (storage version < cluster version). For now it should be enough to depend on N+2 development cycle (pattern used in developing new features where we delay enabling feature to add proto support in separate release, N+1 release adds proto for feature, N+2 enabled the feature).

serathius commented on server/storage/schema/migration.go:
```diff
@@ -0,0 +1,149 @@
+			// execute downgrades in revered order
+			err = s.changes[len(s.changes)-1-i].unsafeDowngrade(tx)
+		}
+		if err != nil {
```
Thanks for catching this. I think we should revert changes to bring us to known state. I have implemented reverts PTAL.

jingyih replied:
This is becoming a very large PR, I will try to finish review in a day or two. Thanks!

serathius commented:
Rewrote code to handle errors during migration. This required me to implement reverts which are much harder to get right. I added a lot of additional tests to make sure that reverts work as expected. 

PTAL @lilic @ptabor @jingyih

serathius commented on server/storage/schema/schema_test.go:
```diff
@@ -20,106 +20,223 @@ import (
-			},
-			expectVersion: &semver.Version{Major: 3, Minor: 6},
+			name:    `V3.6 schema is correct`,
+			version: V3_6,
```
`expectError: false` is default, so question is if we should make it explicit. I think this case is simple enough and clear from the test name that this tests succeeds. Skipping default values is also pretty common pattern in golang. 

I don't think it's needed, but it's not something that I'm attached to.

lilic commented on server/etcdserver/version/monitor.go:
```diff
@@ -73,6 +79,24 @@ func (m *Monitor) UpdateClusterVersionIfNeeded() {
+	if cv == nil {
+		return
+	}
+	m.s.Lock()
```
Do we need this lock for GetStorageVersion? Or could we move to the if function below?

serathius replied:
Here I wanted to make sure that GetStorageVersion and UpdateStorageVersion is done in same transaction.

lilic commented on server/storage/schema/schema.go:
```diff
@@ -18,49 +18,110 @@ import (
 	}
-	copied := V3_5
-	return &copied, nil
+	return V3_5, nil
```
I would still prefer if this did not return a fixed version, but don't really have a better suggestion.

lilic reviewed (commented):
A couple of comments, leaving up to other folks who know the codebase better to approve. But just wanted to say great work!

serathius commented on server/storage/schema/schema.go:
```diff
@@ -18,49 +18,110 @@ import (
 
-// UpdateStorageSchema updates storage version.
-func UpdateStorageSchema(lg *zap.Logger, tx backend.BatchTx) error {
+func init() {
```
Fixed

ptabor commented on server/storage/schema/actions.go:
```diff
@@ -0,0 +1,87 @@
+	return revert, nil
+}
+
+func currentStateFieldAction(tx backend.BatchTx, bucket backend.Bucket, fieldName []byte) action {
```
Please add comment

ptabor replied:
(make it clear that it produces a rollback action)

serathius replied:
I think it's clearer to have in separate function, just the original name was bad. I changed the name to `restoreFieldValueAction`, please let me know what you think.

ptabor commented on server/storage/schema/actions.go:
```diff
@@ -0,0 +1,87 @@
+	return nil
+}
+
+func (as ActionList) unsafeRevert(lg *zap.Logger, tx backend.BatchTx) {
```
unsafeExecuteFromBack (InReverseOrder)

jingyih replied:
+1. maybe also add function description

serathius replied:
Done

jingyih reviewed (commented):
Thanks for the great work! I added few comments. Overall looks good.

serathius commented on server/etcdserver/server.go:
```diff
@@ -2155,6 +2147,19 @@ func (s *EtcdServer) monitorVersions() {
+		case <-s.stopping:
+			return
+		}
+		monitor.UpdateStorageVersionIfNeeded()
```
We usually introduce flags to prevent enabling non-backward compatible changes. For example adding a new wal entry that cannot be read by older version. This is not this case as all changes here are backward compatible. Storage Value field will just be ignored by older version. 

As for migration, it is called to update storage version to be up to date with current binary version. It only implements upgrades, and has same safeguards as cluster version for user attempting downgrade. Current implementation can only upgrade the storage after cluster version was changed  (meaning that all cluster members were already upgraded). This is the safe way to maintain storage version as users will do backup before before upgrading any members.

If we would consider introducing feature flags I would propose to put flag to hide downgrade feature. Current downgrade logic doesn't work at all and can lead to breaking cluster. This way storage version will be updated during upgrades, but downgrading will require explicit flag passed.

jingyih replied:
This make sense. Thanks!

serathius commented on server/storage/schema/actions_test.go:
```diff
@@ -0,0 +1,170 @@
+			},
+			expectState: map[string]string{},
+			expectError: errBrokenAction,
+		},
```
There are tests in migration_test.go, they test migrationStep which wraps ActionList. Those tests mock single actions and validate ordering during revert.

jingyih replied:
Got it. Looks good.

jingyih commented on server/storage/schema/actions.go:
```diff
@@ -0,0 +1,93 @@
+	return nil
+}
+
+// unsafeExecuteInReverseOrder executes actions in revered order. Will panic on
```
Nit: s/revered/reversed

serathius replied:
Done

jingyih reviewed (approved):
LGTM. Thanks again!

serathius commented on server/storage/schema/migration.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+func buildPlan(current semver.Version, target semver.Version) (plan migrationPlan, err error) {
+	for current.Minor != target.Minor {
```
There is no trimToMinor function in semver. I introduce my own version and made version comparison cleaner.

serathius commented on server/storage/schema/migration.go:
```diff
@@ -0,0 +1,119 @@
+		return nil, fmt.Errorf("Changing major storage version is not supported")
+	}
+	// TODO(serathius): Implement downgrades
+	if current.Minor > target.Minor {
```
Done

serathius commented:
Latest changes have broken integrations tests in a way that I don't understand nor can reproduce locally. Reverting to state from yesterday.
